man/mktape.1
.TH mktape "1" "December 2013" "mhvtl 1.5" "User Commands"
.SH NAME
mktape \- Utility to create new/blank virtual media for the VTL.
.SH SYNOPSIS
.B mktape
.B [ \-h ]
.B [ \fIoptions \fR]
.B ...
.SH DESCRIPTION
.\" Add any additional description here
.PP
Create and initialise new media for the Virtual Tape Library. Media is
created in the /opt/mhvtl/ directory by default. This can be over-ridden
using the 'Home directory: ' setting in device.conf.
.SH OPTIONS
.TP
\fB\-h\fR
display this help and exit
.TP
\fB\-l n\fR
Where n is the library index number (default library index numbers 10 & 30).
The 'Home directory: ' entry for the library specified is used as parent
directory for media created.
If the library index number does not exist, then a subdir under /opt/mhvtl/<n>
is created and virtual media created in this directory.
e.g. "vtlcmd -l 20 -m ABC123 -s 1024 -t data -d LTO3" will create a virtual
media 'ABC123' in /opt/mhvtl/20/ (if no library index 20 defined in device.conf)
.TP
\fB\-m PCL\fR
where PCL is the Physical Cartridge Label (barcode).
.TP
\fB\-s size\fR
where size is the capacity of the virtual media - Size is defined in 'megabytes'.
.TP
\fB\-t type\fR
The media 'type' which can be: "data", "WORM" (Write Once Read Many),
"clean" (cleaning cartridge) or "NULL" for media which does not save data.

The NULL is a special media type used primarily to test performance with disk
writes kept to a minimum. Note: Metadata (512bytes per data block) is still
written.
.TP
\fB\-d density\fR
Media density. Can be one of LTO1, LTO2, LTO3, LTO4, SDLT1, SDLT2, SDLT3, SDLT4, AIT1, AIT2, AIT3,
AIT4, T10KA, T10Kb, T10KC, 9840A, 9840B, 9840C, 9840D, 9940A, 9940B, J1A, E05 and E06

The 'J1A, E05 and E06' media densities refer to the IBM 03592 media types.

.SH AUTHOR
Written by Mark Harvey
.SH BUGS
Needs to be made user friendly.
.SH "REPORTING BUGS"
Report bugs to <markh794@gmail.com> <mark_harvey@symantec.com>.
.SH COPYRIGHT
Copyright \(co 2005 Free Software Foundation, Inc.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.BR build_library_config(1),
.BR make_vtl_media(1),
.BR library_contents(5),
.BR mhvtl(1),
.BR vtlcmd(1),
.BR vtllibrary(1),
.BR vtltape(1)
.BR edit_tape(1)


ChangeLog
* Sun Oct 20 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-10
- Various cleanups due to static code check 'smatch'
- Fix TapeAlert bit offsets
- Return bit/byte offset for sense INVALID FIELD IN CDB/PARAMS
- Fix returned data size when 0 data is requested (kernel module fix)
- SSC - Fix op code 0Fh (read reverse) incorrectly set to 'reserve'
- SMC - Ability to keep library media change persistant across restarts
- SSC - Implement 'read media serial number' op code
- Fix 'make_vtl_media' script to understand IBM 03592 media type

* Thu Aug 29 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-9
- SSC -> MODE SELECT overhaul.
- Sanitize delay values introduced in 1.4-8
- daemons now open /dev/mhvtlXX nodes in exclusive mode
- Code cleanup based on Valgrind - memory leaks and unused file handles
- Correctly set SCSI revision field for T10000B
- SMC -> Adjust behaviour if 'Scalar' library defined
 set 'BarC' field in inquiry page and
 pad "Full Firmware Revision" field with compile time/date

* Sat Jun 29 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-8
- Add additional fields in MAM for multi-partition support capability
- MODE SELECT fixes
- Solaris 11.1 st mode select
- Make virtual media more resilient against disk full conditions
- Add additional error checking between kernel and userspace ioctl
- Add ability to introduce delays in load/unload/position/thread/rewind op codes
- Add 16byte CDB locate command support
- Minor kernel module tweeks - increase timeout & error checking on ioctl()

* Fri Mar 22 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-7
- Add DLT7000/DLT8000 emulation
- Record 'last load' data in MAM
- New utility 'edit_tape' which will allow modificiation of media 'meta data'
- Prevent and log possibility of READ ELEMENT STATUS overflowing data buffer
- Set 'POWERON-RESET' if library re-configured via HUP signal
- Update 'minor' number from 'unsigned char' to 'unsigned int' which removes
 max number of device nodes at 256
- Rework READ ELEMENT STATUS - Simplify code and correct bug where incorrect
 byte count returned.
- Add 'support' for specifying 9x40 and DLT media via barcode suffix
- Update kernel module to prevent overflow if DEF_MAX_MINOR_NO reached.
- Update kernel module to allow up to 1024 minor numbers
- Update kernel module vers to 0.18.12, vers date 20130310-0
- off-by-one fix of compression mode page
- Fix 'READ POSITION' ssc op code
- Helper scripts for NetBackup added
 - drive_stats.pl (parse 'scsi_command' output and email on tape unload)
 - vlt_endeject_notify.pl (NetBackup Vault helper)

* Sat Jan 31 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-6
- Return correct TAPE CAPACITY log page
- Fix overflow TAPE CAPACITY log page
- Fix potential vtltape segfault when 'block position back'
- New utility 'edit_tape' to update virtual media metadata

* Sat Jan 12 2013 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-5
- Added support for LTO-6
- Fix EOM handling
- Fix data returned in TAPE CAPACITY log page

* Thu Sep 13 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-4
- Fix a build issue on older (SLES9) linux distributions
- Silence a compiler warning about uninitialised var 'month' in quantum PM
- Note: 1.4-2 & 1.4-3 never made it as released packages.

* Wed Sep 12 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-3
- Add 'performance' counter to vtllibrary
- Fix off-by-one error in LOG SELECT
- Implement LOG reset via PCR bit in LOG SELECT

* Sat Sep 01 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-2
- Add 'performance' counter to each SCSI op code logged

* Wed Aug 08 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-1
- Bug fix:
  Fix dump_tape to understand new library subdirectory format
  Fix vtlcmd to understand new library subdirectory format

* Wed Aug 01 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.4-0
- Allow 'default' personality module to load all media R/W
- Add IBM half-hight INQUIRY string match
- vtltape: Update log entry regarding media capacity depending on capabilities
- vtltape: Make the backoff algrithm value configurable
- Add SDLT600 & SDLT320 personality module
- mhvtl.ko: Silence install 'depmod' error
- Move standard inquiry info into memory structure
- Default media belonging to a library in a subdir
- Remove kmod-mhvtl.spec
- Rename 'mhvtl' package as 'mhvtl-utils'
- Use ELRepo spec file
- Move 'make install' from RPM to Makefile

* Fri Jun 15 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.3-0
- Move 'on disk' unique functions into its own shared lib (libvtlcart.so)
- HP Ultrium VPD page data update to match HP documentation
- IBM Ultrium VPD page 0xC0 & 0xC1 update to match IBM documentation
- Fix SPIN return status
- Add helper functions for moving media within library
  (in prep for library partition manager)
- Add ability to set/clear APPEND ONLY status from user (vtlcmd)

* Wed Apr 04 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.2-3
- Fix segfault in AIT4 when attempting to access Security Protocol IN/OUT
- No need to fail in virtual media parent directory exists
- Add more descriptive debug messages for log/mode page lookups

* Thu Mar 22 2012 Mark Harvey <markh794@gmail.com> <mark_harvey@symantec.com>
- Bumped version to 1.2-2
- Changed defaults so fifo is not enabled. If fifo not used, it will
  block writing process (i.e. hang daemons)
- Improved MODE SELECT / SENSE to support 'changeable field bitmaps
- Started implementing ALLOW OVERWRITE support (incomplete)

usr/mktape.c
/*
 * mktape
 *
 */

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <syslog.h>
#include <string.h>
#include <time.h>
#include <inttypes.h>
#include <pwd.h>
#include "be_byteshift.h"
#include "list.h"
#include "vtl_common.h"
#include "vtltape.h"
#include "vtllib.h"

#if defined _LARGEFILE64_SOURCE
static void *largefile_support = "large file support";
#else
static void *largefile_support = "No largefile support";
#endif

/* The following variables are needed for the MHVTL_DBG() macro to work. */

char vtl_driver_name[] = "mktape";
int verbose = 0;
int debug = 0;
long my_id = 0;
extern char home_directory[HOME_DIR_PATH_SZ + 1];

static void usage(char *progname)
{
	printf("Usage: %s -l lib -m PCL -s size -t type -d density\n",
					progname);
	printf("       Where 'size' is in Megabytes\n");
	printf("             'type' is data | clean | WORM | NULL\n");
	printf("             'PCL' is Physical Cartridge Label (barcode)\n");
	printf("             'lib' is Library number\n");
	printf("             'density' can be on of the following:\n");
	printf("           AIT1     AIT2     AIT3     AIT4\n");
	printf("           DDS1     DDS2     DDS3     DDS4\n");
	printf("           DLT3     DLT4\n");
	printf("           SDLT1    SDLT220  SDLT320  SDLT600\n");
	printf("           LTO1     LTO2     LTO3     LTO4\n");
	printf("           LTO5     LTO6\n");
	printf("           T10KA    T10KB    T10KC\n");
	printf("           9840A    9840B    9840C    9840D\n");
	printf("           9940A    9940B\n");
	printf("           J1A      E05      E06\n\n");
}

int main(int argc, char *argv[])
{
	unsigned char sam_stat;
	char *progname = argv[0];
	char *pcl = NULL;
	char *mediaType = NULL;
	char *mediaCapacity = NULL;
	char *density = NULL;
	char *lib = NULL;
	uint64_t size;
	int libno;
	struct stat statb;
	struct passwd *pw;

	if (sizeof(struct MAM) != 1024) {
		printf("Structure of MAM incorrect size: %d\n",
						(int)sizeof(struct MAM));
		exit(2);
	}

	if (argc < 2) {
		usage(progname);
		exit(1);
	}

	while (argc > 0) {
		if (argv[0][0] == '-') {
			switch (argv[0][1]) {
			case 'd':
				if (argc > 1)
					density = argv[1];
				break;
			case 'l':
				if (argc > 1) {
					lib = argv[1];
				} else {
					puts("    More args needed for -l\n");
					exit(1);
				}
				break;
			case 'm':
				if (argc > 1) {
					pcl = argv[1];
				} else {
					puts("    More args needed for -m\n");
					exit(1);
				}
				break;
			case 's':
				if (argc > 1) {
					mediaCapacity = argv[1];
				} else {
					puts("    More args needed for -s\n");
					exit(1);
				}
				break;
			case 't':
				if (argc > 1) {
					mediaType = argv[1];
				} else {
					puts("    More args needed for -t\n");
					exit(1);
				}
				break;
			case 'V':
				printf("%s: version %s\n%s\n\n",
						progname, MHVTL_VERSION,
						(char *)largefile_support);
				break;
			case 'v':
				verbose++;
				break;
			}
		}
		argv++;
		argc--;
	}

	if (pcl == NULL) {
		printf("Please supply a barcode (-b barcode)\n\n");
		usage(progname);
		exit(1);
	}
	if (mediaCapacity == NULL) {
		printf("Please supply media capacity (-s xx)\n\n");
		usage(progname);
		exit(1);
	}
	if (mediaType == NULL) {
		printf("Please supply cart type (-t data|clean|WORM|NULL)\n\n");
		usage(progname);
		exit(1);
	}
	if (lib == NULL) {
		printf("Please supply Library number (-l xx)\n\n");
		usage(progname);
		exit(1);
	}
	if (density == NULL) {
		printf("Please supply media density (-d xx)\n\n");
		usage(progname);
		exit(1);
	}

	sscanf(mediaCapacity, "%" PRId64, &size);
	if (size == 0)
		size = 8000;

	sscanf(lib, "%d", &libno);
	if (!libno) {
		printf("Invalid library number\n");
		exit(1);
	}

	find_media_home_directory(home_directory, libno);

	if (strlen(pcl) > MAX_BARCODE_LEN) {
		printf("Max barcode length (%d) exceeded\n\n", MAX_BARCODE_LEN);
		usage(progname);
		exit(1);
	}

	pw = getpwnam(USR);	/* Find UID for user 'vtl' */

	/* Verify that the MHVTL home directory exists. */
	if (stat(home_directory, &statb) < 0 && errno == ENOENT) {
		umask(0007);
		if (mkdir(home_directory, 02770) < 0) {
			printf("Cannot create PCL %s, directory %s:"
				"Doesn't exist and cannot be created\n",
						pcl, home_directory);
			exit(1);
		}
	}

	/* Don't really care if this fails or not..
	 * But lets try anyway
	 */
	if (chown(home_directory, pw->pw_uid, pw->pw_gid))
		;

	/* Initialize the contents of the MAM to be used for the new PCL. */
	memset((uint8_t *)&mam, 0, sizeof(mam));

	mam.tape_fmt_version = TAPE_FMT_VERSION;
	mam.mam_fmt_version = MAM_VERSION;
	put_unaligned_be64(size * 1048576, &mam.max_capacity);
	put_unaligned_be64(size * 1048576, &mam.remaining_capacity);
	put_unaligned_be64(sizeof(mam.pad), &mam.MAMSpaceRemaining);

	memcpy(&mam.MediumManufacturer, "linuxVTL", 8);
	memcpy(&mam.ApplicationVendor, "vtl-1.4 ", 8);
	sprintf((char *)mam.ApplicationVersion, "%d", TAPE_FMT_VERSION);

	if (!strncmp("clean", mediaType, 5)) {
		mam.MediumType = MEDIA_TYPE_CLEAN;	/* Cleaning cart */
		mam.MediumTypeInformation = 20;		/* Max cleaning loads */
	} else if (!strncmp("NULL", mediaType, 4)) {
		mam.MediumType = MEDIA_TYPE_NULL;	/* save metadata only */
	} else if (!strncmp("WORM", mediaType, 4)) {
		mam.MediumType = MEDIA_TYPE_WORM;	/* WORM cart */
	} else {
		mam.MediumType = MEDIA_TYPE_DATA;	/* Normal data cart */
	}
	set_media_params(&mam, density);

	sprintf((char *)mam.MediumSerialNumber, "%s_%d", pcl, (int)time(NULL));
	sprintf((char *)mam.MediumManufactureDate, "%d", (int)time(NULL));
	sprintf((char *)mam.Barcode, "%-31s", pcl);

	/* Create the PCL using the initialized MAM. */

	exit(create_tape(pcl, &mam, &sam_stat));
}
